File: frameworks/flutter/bota_flutter_sdk/ios/bota_flutter_sdk/Sources/bota_flutter_sdk/NativeLeaseCoordinator.swift
import Foundation

protocol NativeLeaseClientProtocol: Sendable {
  func configure(applicationSupportDirectory: URL) async throws
  func destroy() async
}

struct NativeLeaseConfiguration: Equatable, Sendable {
  let applicationSupportDirectory: URL
  let applicationSupportNamespace: String
  let hasProvisioningMaterialCallback: Bool
  let hasFactoryResetGrantCallback: Bool
  let hasFactoryResetResultCallback: Bool
  let hasFirmwareCallback: Bool
}

enum NativeOperationCategory: CaseIterable, Hashable, Sendable {
  case device
  case provisioning
  case factoryReset
  case recording
  case ota
  case logs
  case wifi
}

enum NativeLeaseError: Error, Equatable {
  case configurationConflict
  case acquisitionCancelled
  case operationInProgress
  case operationNotOwned
}

struct NativeOwnedOperation: Equatable, Sendable {
  let category: NativeOperationCategory
  let operationID: String
}

actor NativeLeaseCoordinator {
  static let shared = NativeLeaseCoordinator(client: BotaAppleNativeClient.shared)

  private struct ConfigurationTask {
    let id: UUID
    let configuration: NativeLeaseConfiguration
    let task: Task<Void, Error>
  }

  private struct DestroyTask {
    let id: UUID
    let task: Task<Void, Never>
  }

  private enum CancellationOutcome: @unchecked Sendable {
    case success
    case failure(Error)
  }

  private struct OperationCancellation {
    let id: UUID
    let task: Task<CancellationOutcome, Never>
    var status: CancellationStatus
  }

  private enum CancellationStatus: Equatable {
    case inProgress
    case succeeded
    case failed
  }

  private struct OperationOwner {
    let engineID: String
    let operationID: String
    var cancellation: OperationCancellation?
    var operationTerminated: Bool
    var retainAfterCancellation: Bool
  }

  private let client: any NativeLeaseClientProtocol
  private var leases: [String: NativeLeaseConfiguration] = [:]
  private var pendingLeases: [String: NativeLeaseConfiguration] = [:]
  private var activeConfiguration: NativeLeaseConfiguration?
  private var configuring: ConfigurationTask?
  private var destroying: DestroyTask?
  private var operations: [NativeOperationCategory: OperationOwner] = [:]

  init(client: any NativeLeaseClientProtocol) {
    self.client = client
  }

  var leaseCount: Int { leases.count }

  func acquire(engineID: String, configuration: NativeLeaseConfiguration) async throws {
    if let existing = leases[engineID] {
      guard existing == configuration else { throw NativeLeaseError.configurationConflict }
      return
    }
    if let pending = pendingLeases[engineID] {
      guard pending == configuration else { throw NativeLeaseError.configurationConflict }
    } else {
      try requireCompatible(configuration)
      pendingLeases[engineID] = configuration
    }

    if let destroying {
      await destroying.task.value
      if self.destroying?.id == destroying.id { self.destroying = nil }
      guard pendingLeases[engineID] != nil else {
        throw NativeLeaseError.acquisitionCancelled
      }
    }

    if let activeConfiguration {
      guard activeConfiguration == configuration else {
        pendingLeases.removeValue(forKey: engineID)
        throw NativeLeaseError.configurationConflict
      }
    } else {
      let configurationTask: ConfigurationTask
      if let configuring {
        guard configuring.configuration == configuration else {
          pendingLeases.removeValue(forKey: engineID)
          throw NativeLeaseError.configurationConflict
        }
        configurationTask = configuring
      } else {
        let client = self.client
        configurationTask = ConfigurationTask(
          id: UUID(),
          configuration: configuration,
          task: Task {
            try await client.configure(
              applicationSupportDirectory: configuration.applicationSupportDirectory
            )
          }
        )
        configuring = configurationTask
      }

      do {
        try await configurationTask.task.value
        completeConfiguration(configurationTask)
      } catch {
        if configuring?.id == configurationTask.id { configuring = nil }
        pendingLeases.removeValue(forKey: engineID)
        throw error
      }
    }

    if let existing = leases[engineID] {
      guard existing == configuration else { throw NativeLeaseError.configurationConflict }
      return
    }
    guard pendingLeases.removeValue(forKey: engineID) != nil else {
      await destroyIfUnused()
      throw NativeLeaseError.acquisitionCancelled
    }
    leases[engineID] = configuration
  }

  func release(engineID: String) async {
    let removedLease = leases.removeValue(forKey: engineID)
    let removedPending = pendingLeases.removeValue(forKey: engineID)
    guard removedLease != nil || removedPending != nil else { return }
    guard leases.isEmpty, pendingLeases.isEmpty else { return }

    // The suspended acquirer owns completion cleanup. Detach must not wait for
    // native configuration to return before it can revoke the pending lease.
    if configuring != nil { return }
    await destroyIfUnused()
  }

  func beginOperation(
    engineID: String,
    category: NativeOperationCategory,
    operationID: String
  ) throws {
    guard operations[category] == nil else { throw NativeLeaseError.operationInProgress }
    operations[category] = OperationOwner(
      engineID: engineID,
      operationID: operationID,
      cancellation: nil,
      operationTerminated: false,
      retainAfterCancellation: false
    )
  }

  func finishOperation(
    engineID: String,
    category: NativeOperationCategory,
    operationID: String
  ) {
    guard var owner = operations[category],
      owner.engineID == engineID,
      owner.operationID == operationID
    else { return }
    owner.operationTerminated = true
    guard let cancellation = owner.cancellation else {
      if owner.retainAfterCancellation {
        operations[category] = owner
      } else {
        operations.removeValue(forKey: category)
      }
      return
    }
    if cancellation.status == .inProgress || owner.retainAfterCancellation {
      operations[category] = owner
    } else {
      operations.removeValue(forKey: category)
    }
  }

  func cancelOperation(
    engineID: String,
    category: NativeOperationCategory,
    cancellation: @escaping @Sendable () async throws -> Void
  ) async throws {
    guard var owner = operations[category] else { return }
    guard owner.engineID == engineID else { throw NativeLeaseError.operationNotOwned }
    guard owner.cancellation == nil else { throw NativeLeaseError.operationInProgress }

    let cancellationID = UUID()
    let task = Task<CancellationOutcome, Never> {
      do {
        try await cancellation()
        return .success
      } catch {
        return .failure(error)
      }
    }
    owner.cancellation = OperationCancellation(
      id: cancellationID,
      task: task,
      status: .inProgress
    )
    operations[category] = owner
    let outcome = await task.value
    settleCancellation(
      engineID: engineID,
      category: category,
      operationID: owner.operationID,
      cancellationID: cancellationID,
      outcome: outcome
    )
    if case .failure(let error) = outcome { throw error }
  }

  func beginCancellingOperations(
    engineID: String,
    cancellation: @escaping @Sendable (NativeOperationCategory) async throws -> Void
  ) -> [NativeOwnedOperation] {
    let owned: [NativeOwnedOperation] = NativeOperationCategory.allCases.compactMap {
      category -> NativeOwnedOperation? in
      guard let owner = operations[category], owner.engineID == engineID else { return nil }
      return NativeOwnedOperation(category: category, operationID: owner.operationID)
    }
    for operation in owned {
      guard var owner = operations[operation.category],
        owner.engineID == engineID,
        owner.operationID == operation.operationID
      else { continue }
      owner.retainAfterCancellation = true
      if owner.cancellation == nil {
        let category = operation.category
        owner.cancellation = OperationCancellation(
          id: UUID(),
          task: Task {
            do {
              try await cancellation(category)
              return .success
            } catch {
              return .failure(error)
            }
          },
          status: .inProgress
        )
      }
      operations[operation.category] = owner
    }
    return owned
  }

  func waitForOperationCancellations(
    engineID: String,
    operations owned: [NativeOwnedOperation]
  ) async {
    for operation in owned {
      guard let owner = operations[operation.category],
        owner.engineID == engineID,
        owner.operationID == operation.operationID,
        let cancellation = owner.cancellation
      else { continue }
      let outcome = await cancellation.task.value
      settleCancellation(
        engineID: engineID,
        category: operation.category,
        operationID: operation.operationID,
        cancellationID: cancellation.id,
        outcome: outcome
      )
    }
  }

  func finishCancelledOperations(
    engineID: String,
    operations owned: [NativeOwnedOperation],
    terminalOperationIDs: Set<String>
  ) {
    for operation in owned {
      guard var owner = operations[operation.category],
        owner.engineID == engineID,
        owner.operationID == operation.operationID,
        owner.retainAfterCancellation
      else { continue }
      if terminalOperationIDs.contains(operation.operationID) {
        owner.operationTerminated = true
      }
      owner.retainAfterCancellation = false
      guard let cancellation = owner.cancellation else {
        if owner.operationTerminated {
          operations.removeValue(forKey: operation.category)
        } else {
          operations[operation.category] = owner
        }
        continue
      }
      if cancellation.status == .succeeded || owner.operationTerminated {
        operations.removeValue(forKey: operation.category)
      } else {
        operations[operation.category] = owner
      }
    }
  }

  private func settleCancellation(
    engineID: String,
    category: NativeOperationCategory,
    operationID: String,
    cancellationID: UUID,
    outcome: CancellationOutcome
  ) {
    guard var owner = operations[category],
      owner.engineID == engineID,
      owner.operationID == operationID,
      var cancellation = owner.cancellation,
      cancellation.id == cancellationID
    else { return }
    switch outcome {
    case .success: cancellation.status = .succeeded
    case .failure: cancellation.status = .failed
    }
    owner.cancellation = cancellation
    let nativeCleanupSucceeded = cancellation.status == .succeeded
    if !owner.retainAfterCancellation && (nativeCleanupSucceeded || owner.operationTerminated) {
      operations.removeValue(forKey: category)
    } else {
      operations[category] = owner
    }
  }

  private func requireCompatible(_ configuration: NativeLeaseConfiguration) throws {
    let current =
      activeConfiguration
      ?? configuring?.configuration
      ?? leases.values.first
      ?? pendingLeases.values.first
    if let current, current != configuration { throw NativeLeaseError.configurationConflict }
  }

  private func completeConfiguration(_ completed: ConfigurationTask) {
    guard configuring?.id == completed.id else { return }
    activeConfiguration = completed.configuration
    configuring = nil
  }

  private func destroyIfUnused() async {
    guard leases.isEmpty, pendingLeases.isEmpty else { return }
    if let destroying {
      await destroying.task.value
      if self.destroying?.id == destroying.id { self.destroying = nil }
      operations.removeAll()
      return
    }
    guard activeConfiguration != nil else {
      if configuring == nil { operations.removeAll() }
      return
    }
    activeConfiguration = nil
    let client = self.client
    let destroy = DestroyTask(id: UUID(), task: Task { await client.destroy() })
    destroying = destroy
    await destroy.task.value
    if destroying?.id == destroy.id { destroying = nil }
    operations.removeAll()
  }
}

File: frameworks/flutter/bota_flutter_sdk/android/src/main/kotlin/dev/bota/sdk/flutter/NativeLeaseCoordinator.kt
package dev.bota.sdk.flutter

import java.util.UUID
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

internal interface NativeLeaseClient {
    suspend fun configure(namespace: String)
    suspend fun destroy()
}

internal data class NativeLeaseConfiguration(
    val applicationSupportNamespace: String,
    val hasProvisioningMaterialCallback: Boolean,
    val hasFactoryResetGrantCallback: Boolean,
    val hasFactoryResetResultCallback: Boolean,
    val hasFirmwareCallback: Boolean,
)

internal enum class NativeOperationCategory {
    DEVICE,
    PROVISIONING,
    FACTORY_RESET,
    RECORDING,
    OTA,
    LOGS,
    WIFI,
}

internal sealed class NativeLeaseException(message: String) : IllegalStateException(message) {
    class ConfigurationConflict : NativeLeaseException("native client is configured differently")
    class AcquisitionCancelled : NativeLeaseException("native lease acquisition was cancelled")
    class OperationInProgress : NativeLeaseException("another engine owns this native operation")
    class OperationNotOwned : NativeLeaseException("native operation belongs to another engine")
}

internal data class NativeOwnedOperation(
    val category: NativeOperationCategory,
    val operationId: String,
)

internal class NativeLeaseCoordinator(
    private val client: NativeLeaseClient,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate),
    private val beforeOperationAcquisition: suspend () -> Unit = {},
) {
    private data class ConfigurationAttempt(
        val id: UUID,
        val configuration: NativeLeaseConfiguration,
        val task: Deferred<Result<Unit>>,
    )

    private data class DestroyAttempt(
        val id: UUID,
        val task: Deferred<Unit>,
    )

    private enum class CancellationStatus { IN_PROGRESS, SUCCEEDED, FAILED }

    private data class OperationCancellation(
        val id: UUID,
        val task: Deferred<Result<Unit>>,
        var status: CancellationStatus,
    )

    private data class OperationOwner(
        val engineId: String,
        val operationId: String,
        var cancellation: OperationCancellation? = null,
        var operationTerminated: Boolean = false,
        var retainAfterCancellation: Boolean = false,
    )

    private val lock = Mutex()
    private val leases = mutableMapOf<String, NativeLeaseConfiguration>()
    private val pendingLeases = mutableMapOf<String, NativeLeaseConfiguration>()
    private val operations = mutableMapOf<NativeOperationCategory, OperationOwner>()
    private var activeConfiguration: NativeLeaseConfiguration? = null
    private var configuring: ConfigurationAttempt? = null
    private var destroying: DestroyAttempt? = null

    suspend fun acquire(engineId: String, configuration: NativeLeaseConfiguration) {
        var destroyToWait: DestroyAttempt? = null
        var configurationAttempt: ConfigurationAttempt? = null
        lock.withLock {
            leases[engineId]?.let {
                if (it != configuration) throw NativeLeaseException.ConfigurationConflict()
                return
            }
            pendingLeases[engineId]?.let {
                if (it != configuration) throw NativeLeaseException.ConfigurationConflict()
            } ?: run {
                requireCompatible(configuration)
                pendingLeases[engineId] = configuration
            }
            destroyToWait = destroying
            if (destroyToWait == null) {
                activeConfiguration?.let {
                    if (it != configuration) {
                        pendingLeases.remove(engineId)
                        throw NativeLeaseException.ConfigurationConflict()
                    }
                } ?: run {
                    configurationAttempt = configurationAttempt(configuration)
                }
            }
        }

        destroyToWait?.task?.await()
        if (destroyToWait != null) {
            lock.withLock {
                if (destroying?.id == destroyToWait.id) destroying = null
                if (!pendingLeases.containsKey(engineId)) {
                    throw NativeLeaseException.AcquisitionCancelled()
                }
                activeConfiguration?.let {
                    if (it != configuration) {
                        pendingLeases.remove(engineId)
                        throw NativeLeaseException.ConfigurationConflict()
                    }
                } ?: run {
                    configurationAttempt = configurationAttempt(configuration)
                }
            }
        }

        configurationAttempt?.let { attempt ->
            val outcome = attempt.task.await()
            outcome.exceptionOrNull()?.let { error ->
                lock.withLock {
                    if (configuring?.id == attempt.id) configuring = null
                    pendingLeases.remove(engineId)
                }
                throw error
            }
            lock.withLock {
                if (configuring?.id == attempt.id) {
                    activeConfiguration = attempt.configuration
                    configuring = null
                }
            }
        }

        var cancelled = false
        lock.withLock {
            leases[engineId]?.let {
                if (it != configuration) throw NativeLeaseException.ConfigurationConflict()
                return
            }
            if (pendingLeases.remove(engineId) == null) {
                cancelled = true
            } else {
                leases[engineId] = configuration
            }
        }
        if (cancelled) {
            destroyIfUnused()
            throw NativeLeaseException.AcquisitionCancelled()
        }
    }

    suspend fun release(engineId: String) {
        val shouldDestroy = lock.withLock {
            val removed = leases.remove(engineId) != null || pendingLeases.remove(engineId) != null
            removed && leases.isEmpty() && pendingLeases.isEmpty() && configuring == null
        }
        if (shouldDestroy) destroyIfUnused()
    }

    suspend fun beginOperation(
        engineId: String,
        category: NativeOperationCategory,
        operationId: String,
    ) {
        beforeOperationAcquisition()
        lock.withLock {
            if (operations[category] != null) throw NativeLeaseException.OperationInProgress()
            operations[category] = OperationOwner(engineId, operationId)
        }
    }

    suspend fun finishOperation(
        engineId: String,
        category: NativeOperationCategory,
        operationId: String,
    ) {
        lock.withLock {
            val owner = operations[category] ?: return
            if (owner.engineId != engineId || owner.operationId != operationId) return
            owner.operationTerminated = true
            val cancellation = owner.cancellation
            if (!owner.retainAfterCancellation &&
                (cancellation == null || cancellation.status != CancellationStatus.IN_PROGRESS)
            ) {
                operations.remove(category)
            }
        }
    }

    suspend fun cancelOperation(
        engineId: String,
        category: NativeOperationCategory,
        cancellation: suspend () -> Unit,
    ) {
        val ownerAndCancellation = lock.withLock {
            val owner = operations[category] ?: return
            if (owner.engineId != engineId) throw NativeLeaseException.OperationNotOwned()
            if (owner.cancellation != null) throw NativeLeaseException.OperationInProgress()
            val operationCancellation = OperationCancellation(
                id = UUID.randomUUID(),
                task = scope.async { runCatching { cancellation() } },
                status = CancellationStatus.IN_PROGRESS,
            )
            owner.cancellation = operationCancellation
            owner to operationCancellation
        }
        val (owner, operationCancellation) = ownerAndCancellation
        val outcome = operationCancellation.task.await()
        settleCancellation(owner, category, operationCancellation.id, outcome)
        outcome.getOrThrow()
    }

    suspend fun beginCancellingOperations(
        engineId: String,
        cancellation: suspend (NativeOperationCategory) -> Unit,
    ): List<NativeOwnedOperation> = lock.withLock {
        val owned = NativeOperationCategory.entries.mapNotNull { category ->
            operations[category]?.takeIf { it.engineId == engineId }?.let {
                NativeOwnedOperation(category, it.operationId)
            }
        }
        owned.forEach { operation ->
            val owner = operations[operation.category] ?: return@forEach
            owner.retainAfterCancellation = true
            if (owner.cancellation == null) {
                owner.cancellation = OperationCancellation(
                    id = UUID.randomUUID(),
                    task = scope.async { runCatching { cancellation(operation.category) } },
                    status = CancellationStatus.IN_PROGRESS,
                )
            }
        }
        owned
    }

    suspend fun waitForOperationCancellations(
        engineId: String,
        owned: List<NativeOwnedOperation>,
    ) {
        owned.forEach { operation ->
            val cancellation = lock.withLock {
                operations[operation.category]
                    ?.takeIf { it.engineId == engineId && it.operationId == operation.operationId }
                    ?.cancellation
            } ?: return@forEach
            val outcome = cancellation.task.await()
            val owner = lock.withLock { operations[operation.category] } ?: return@forEach
            settleCancellation(owner, operation.category, cancellation.id, outcome)
        }
    }

    suspend fun finishCancelledOperations(
        engineId: String,
        owned: List<NativeOwnedOperation>,
        terminalOperationIds: Set<String>,
    ) {
        lock.withLock {
            owned.forEach { operation ->
                val owner = operations[operation.category]
                    ?.takeIf {
                        it.engineId == engineId &&
                            it.operationId == operation.operationId &&
                            it.retainAfterCancellation
                    } ?: return@forEach
                if (operation.operationId in terminalOperationIds) owner.operationTerminated = true
                owner.retainAfterCancellation = false
                val cancellationSucceeded = owner.cancellation?.status == CancellationStatus.SUCCEEDED
                if (cancellationSucceeded || owner.operationTerminated) {
                    operations.remove(operation.category)
                }
            }
        }
    }

    private fun configurationAttempt(configuration: NativeLeaseConfiguration): ConfigurationAttempt {
        configuring?.let {
            if (it.configuration != configuration) throw NativeLeaseException.ConfigurationConflict()
            return it
        }
        return ConfigurationAttempt(
            id = UUID.randomUUID(),
            configuration = configuration,
            task = scope.async { runCatching { client.configure(configuration.applicationSupportNamespace) } },
        ).also { configuring = it }
    }

    private fun requireCompatible(configuration: NativeLeaseConfiguration) {
        val current = activeConfiguration
            ?: configuring?.configuration
            ?: leases.values.firstOrNull()
            ?: pendingLeases.values.firstOrNull()
        if (current != null && current != configuration) {
            throw NativeLeaseException.ConfigurationConflict()
        }
    }

    private suspend fun settleCancellation(
        expected: OperationOwner,
        category: NativeOperationCategory,
        cancellationId: UUID,
        outcome: Result<Unit>,
    ) {
        lock.withLock {
            val owner = operations[category] ?: return
            if (owner.engineId != expected.engineId || owner.operationId != expected.operationId) return
            val cancellation = owner.cancellation?.takeIf { it.id == cancellationId } ?: return
            cancellation.status = if (outcome.isSuccess) {
                CancellationStatus.SUCCEEDED
            } else {
                CancellationStatus.FAILED
            }
            if (!owner.retainAfterCancellation &&
                (cancellation.status == CancellationStatus.SUCCEEDED || owner.operationTerminated)
            ) {
                operations.remove(category)
            }
        }
    }

    private suspend fun destroyIfUnused() {
        val attempt = lock.withLock {
            if (leases.isNotEmpty() || pendingLeases.isNotEmpty()) return
            destroying?.let { return@withLock it }
            if (activeConfiguration == null) {
                if (configuring == null) operations.clear()
                return
            }
            activeConfiguration = null
            DestroyAttempt(UUID.randomUUID(), scope.async { client.destroy() }).also { destroying = it }
        }
        attempt.task.await()
        lock.withLock {
            if (destroying?.id == attempt.id) destroying = null
            operations.clear()
        }
    }

    internal companion object {
        val shared: NativeLeaseCoordinator by lazy { NativeLeaseCoordinator(BotaAndroidNativeClient) }
    }
}
